File: akkaActorsPersistent/src/main/scala/com/lightbend/scala/modelserver/actor/actors/ModelServingManager.scala
package com.lightbend.scala.modelserver.actor.actors

import akka.actor.{Actor, ActorRef, Props}
import com.lightbend.model.winerecord.WineRecord
import com.lightbend.scala.modelServer.model.{ModelToServeStats, ModelWithDescriptor}


// Router actor, routing both model and data to an appropriate actor
// Based on http://michalplachta.com/2016/01/23/scalability-using-sharding-from-akka-cluster/

class ModelServingManager extends Actor {

  private def getModelServer(dataType: String): ActorRef = {
    context.child(dataType).getOrElse(context.actorOf(ModelServingActor.props(dataType), dataType))
  }

  private def getInstances : GetModelsResult =
    GetModelsResult(context.children.map(_.path.name).toSeq)

  override def receive = {
    case model: ModelWithDescriptor => getModelServer(model.descriptor.dataType) forward model

    case record: WineRecord => getModelServer(record.dataType) forward record

    case getState: GetState => {
      context.child(getState.dataType) match{
        case Some(server) => server forward getState
        case _ => sender() ! ModelToServeStats.empty
      }
    }

    // Exercise: Provide implementation here.
      // If the actor getState.dataType exists -> forward a request to it.
      // Otherwise return an empty ModelToServeStats:
      // 1. Use the actor context to get the child for the state (`getState.dataType`)
      // 2. Match on the returned value, which will be an Option[ActorRef].
      // 3. If a Some(ref), forward the state to the ref
      // 4. Otherwise, send the empty `ModelToServeStats` as a message to the `sender`.

    case getModels : GetModels => sender() ! getInstances
  }
}

object ModelServingManager{
  def props : Props = Props(new ModelServingManager())
}

case class GetModels()

case class GetModelsResult(models : Seq[String])

File: akkaActorsPersistent/src/main/java/com/lightbend/java/modelserver/actor/actors/ModelServingManager.java
package com.lightbend.java.modelserver.actor.actors;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import com.google.common.collect.Lists;
import com.lightbend.java.model.ModelServingInfo;
import com.lightbend.java.model.ModelWithDescriptor;
import com.lightbend.model.Winerecord;
import scala.Option;

import java.util.LinkedList;
import java.util.List;

public class ModelServingManager extends AbstractActor {
    private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

    private ActorRef getModelServer(String dataType){

        ActorContext context = getContext();
        Option<ActorRef> mayBeRef = getContext().child(dataType);
        return  mayBeRef.isEmpty() ? context.actorOf(ModelServingActor.props(dataType), dataType) : mayBeRef.get();
    }

    private List<String> getInstances() {

        List<String> names = new LinkedList<>();
        List<ActorRef> children = Lists.newArrayList(getContext().getChildren());
        for(ActorRef ref : children)
            names.add(ref.path().name());
        return names;
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(ModelWithDescriptor.class, modelWithDescriptor -> {
                    ActorRef destination = getModelServer(modelWithDescriptor.getDescriptor().getDataType());
                    destination.forward(modelWithDescriptor, getContext());
                })
                .match(Winerecord.WineRecord.class, dataRecord -> {
                    ActorRef destination = getModelServer(dataRecord.getDataType());
                    destination.forward(dataRecord, getContext());
                })
                .match(GetState.class, dataType -> {
                    Option<ActorRef> mayBeRef = getContext().child(dataType.getDataType());
                    if(mayBeRef.isEmpty())
                        getSender().tell(ModelServingInfo.empty, getSelf());
                    else
                        mayBeRef.get().forward(dataType, getContext());
                    })
                .match(GetModels.class, models -> {
                    getSender().tell(getInstances(), getSelf());
                })
                .build();
    }
}
